adoc/chapters/architecture.adoc

```diff
@@ -1099,23 +1098,42 @@ to the buffer or USM allocation from device code.
 
 ==== Forward progress
 
-Memory consistency guarantees are independent of any forward progress
-guarantees.  A SYCL implementation must execute work-items concurrently
-and must ensure that the work-items in a group obey the semantics of
-<<group-barrier,group barriers>>, but are not required to provide any
-additional forward progress guarantees.
+This section, and any subsequent section referring to progress guarantees, uses
+the following terms as defined in the {cpp}17 specification: thread of
+execution; weakly parallel forward progress guarantees; parallel forward
+progress guarantees; concurrent forward progress guarantees; and block with
+forward progress guarantee delegation.
+
+Each work-item in SYCL is a separate thread of execution, providing at least
+weakly parallel forward progress guarantees.  Whether work-items provide
+stronger forward progress guarantees is implementation-defined.
+
+All implementations must additionally ensure that the work-items in a group
+obey the semantics of <<group-barrier,group barriers>>: when a work-item
+arrives at a group barrier acting on group _G_, implementations must eventually
+select and potentially strengthen another work-item in group _G_ that has not
+yet arrived at the barrier.
 
 Synchronizing work-items via memory operations is unsafe in general, but is
 supported if and only if the following conditions are true:
 
   * acquire-release or sequentially consistent memory ordering is supported
     at the scope containing the set of work-items being synchronized;
 
-  * the work-items being synchronized are guaranteed to make forward progress
-    with respect to one another.
+  * the work-items being synchronized both provide concurrent forward progress
+    guarantees.
```

Comment:
Some interpretations of "synchronize" probably don't require concurrent forward progress guarantees.  Should this be strengthened to "synchronizes with", and the term imported from C++ above in [L1101-1105]?(https://github.com/KhronosGroup/SYCL-Docs/pull/300/files#diff-3211f50334dd66e4638c43c11bb1b9dc16f0c26a8b7e5af104ea4e9f7b1744ecR1101-R1105)

Reply:
You're right, this section is unclear.  Reading this again, I think we might even be conflating two concepts.

The first bullet point definitely means "synchronizes with". A device needs to support acquire-release ordering or sequential consistency in order to have atomics from different work-items synchronize with one another.

The second bullet point is a more colloquial usage of "synchronized", and I think it's actually talking about the situation where one work-item attempts to synchronize with another using an atomic operation that is not "lock-free".  It's the blocking nature of the atomic operation that requires strong forward progress guarantees.

I'm not sure what to do here.  If we import "synchronizes with" and "lock-free" we potentially have to rewrite large parts of the specification (e.g. the section currently titled "Synchronization"!).  We might want to do that eventually, but it seems like something we could tackle separately.

Another option would be to drop these bullet points from the specification temporarily and reintroduce them later. Since they're only offering guidance to developers, it may make more sense to have them as non-normative notes in a different section.

Reply:
I've drafted a change in https://github.com/KhronosGroup/SYCL-Docs/pull/300/commits/3a5e79070b21d6a6d4a215838bf2aebbdd9be386 to show what the last option would look like.  I think this fixes the immediate issue of the paragraph you pointed to being confusing.

I'll open an internal issue to discuss "synchronizes-with" further.